File: src/BaseHmiTypes/Screens/Defaults/HmiEffectivePropertyResolver.cs
using BaseHmiTypes.Screens.Base;

namespace BaseHmiTypes.Screens.Defaults;

public sealed class HmiEffectivePropertyResolver
{
    public HmiEffectivePropertyResolver(HmiDefaultProfile? defaultProfile = null)
    {
        DefaultProfile = defaultProfile ?? HmiDefaultProfiles.Empty;
    }

    public HmiDefaultProfile DefaultProfile { get; }

    public HmiProperty<T>? Resolve<T>(HmiScreenItemBase item, string propertyName, HmiProperty<T>? property)
    {
        if (property != null)
            return property;

        return DefaultProfile.TryGet<T>(item, propertyName, out var defaultValue)
            ? HmiProperty.Default(defaultValue, DefaultProfile.Name, propertyName)
            : null;
    }

    public bool TryGetStaticValue<T>(HmiScreenItemBase item, string propertyName, HmiProperty<T>? property, out T value)
    {
        value = default!;
        var resolved = Resolve(item, propertyName, property);
        if (resolved == null)
            return false;

        object? staticValue = resolved.StaticValue;
        if (staticValue == null)
            return false;

        value = (T)staticValue;
        return true;
    }
}

File: typescript/base-hmi-types/src/screens/defaults/HmiEffectivePropertyResolver.ts
import { defaultProperty, HmiProperty } from "../base/HmiProperty.js";
import { HmiScreenItemBase } from "../base/HmiScreenItemBase.js";
import { HmiDefaultProfile } from "./HmiDefaultProfile.js";
import { HmiDefaultProfiles } from "./HmiDefaultProfiles.js";

export class HmiEffectivePropertyResolver {
  constructor(readonly defaultProfile: HmiDefaultProfile = HmiDefaultProfiles.empty) {}

  resolve<T>(item: HmiScreenItemBase, propertyName: string, property: HmiProperty<T> | undefined): HmiProperty<T> | undefined {
    if (property !== undefined) {
      return property;
    }

    const defaultValue = this.defaultProfile.tryGet<T>(item, propertyName);
    return defaultValue.found ? defaultProperty(defaultValue.value, this.defaultProfile.name, propertyName) : undefined;
  }

  tryGetStaticValue<T>(
    item: HmiScreenItemBase,
    propertyName: string,
    property: HmiProperty<T> | undefined,
  ): { found: boolean; value?: T } {
    const resolved = this.resolve(item, propertyName, property);
    if (resolved?.staticValue === undefined || resolved.staticValue === null) {
      return { found: false };
    }

    return { found: true, value: resolved.staticValue };
  }
}
